### PR TITLE
Add monitor theia extension

### PR DESCRIPTION
**Pull request by sgraband**
* Add monitor theia-extension
Added `monitor-theia` extension.
Works the same way as the monitor vscode extension.
Adjusted the way ports are replaced in deployment/service.
This allows to specify the same port for the monitor and the app itself.
Also make sure that the monitor not being enabled does not crash the operator.

* Delete theia-activity-tracker

* Adjust monitor values in value file

Contributed on behalf of STMicroelectronics.

Corresponding helm chart change: https://github.com/eclipsesource/theia-cloud-helm/pull/20

**To test:**
1. Build the operator and service images as `theia-cloud-(operator|service):local`
2. Pull the `simongraband/theia-cloud-demo:local` image. (It contains a plain theia application with the theia extension)
3. Retag the image to `theia-cloud-demo:local`
4. Checkout the helm chart PR
5. If you want you can adjust the monitor values in `valuesMinikubeMonitor.yaml` (e.g. reducing the `notifyAfter` and `timeoutAfter` values)
6. Do `helm install` using the checked out helm chart and the `valuesMinikubeMonitor.yaml` values file
7. See that the activity is tracked in the operator logs and that after your time limits a warning and a pod closed message are displayed to the user.

### CODE REVIEW

**Review by jfaltermeier — CHANGES_REQUESTED** [review]
Thanks, I had a first testing round and had to insert a space in the yaml generation to get it running (see suggestion). With that I was able to test the fixed timeout and the vscode extension. 

When testing the Theia Extension locally I currently get this error in the operator when pinging the last activity:
```
09:03:15.769 [pool-4-thread-1] INFO  org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - [ws-asdfghjkl-theia-cloud-demo-foo-theia-cloud-io-session] Pinging session at 10.107.95.253
09:03:15.771 [pool-4-thread-1] INFO  org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - [ws-asdfghjkl-theia-cloud-demo-foo-theia-cloud-io-session] GET http://10.107.95.253:3000/monitor/activity/lastActivity
09:03:15.786 [pool-4-thread-1] INFO  org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - [ws-asdfghjkl-theia-cloud-demo-foo-theia-cloud-io-session] REQUEST FAILED: GET http://10.107.95.253:3000/monitor/activity/lastActivity. Error: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```
Have you seen this before?

I will now also start having a look at the code.

File: node/monitor-theia/src/common/util.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { THEIACLOUD_SESSION_SECRET } from './env-variables';
+
+export function isAuthorized(req: any): boolean {
```
Comment (jfaltermeier) [inline node/monitor-theia/src/common/util.ts:3]: I have a question about this. This is the auth between monitor and the running sessions and has nothing to do with the user authentication, isn't it? So this is supposed to be authenticated in all cases?

File: node/monitor-theia/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  "keywords": [
+    "theia-extension"
+  ],
+  "version": "0.8.0",
+  "files": [
+    "lib",
+    "src"
+  ],
+  "dependencies": {
+    "@theia/core": "latest",
```
[the remark this answers is not in the document]
Comment (sgraband) [reply node/monitor-theia/package.json:12]: Yes that definatley makes sense. I will just go with the latest community release, then we could easily update it for future community releases as well.

**Comment by sgraband** [comment]
Unfortunately i haven't encountered this error messages before. How does your local tets setup look like?

**Comment by jfaltermeier** [comment]
> Unfortunately i haven't encountered this error message before. How does your local test setup look like?

Basically I was using the terraform chart that is available on main: https://github.com/eclipsesource/theia-cloud/blob/main/terraform/terraform.md#minikube
This does some magic with regards to default certificate, so I will retry this then.

File: node/monitor-theia/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfaltermeier) [reply node/monitor-theia/package.json:12]: Sounds good

**Comment by jfaltermeier** [comment]
> I tried the setup on GKE with valid certificates now and am facing a different issue. Here it seems that the Get call never returns:
> 
> ```
> 08:42:30.584 [pool-5-thread-1] INFO  org.eclipse.theia.cloud.operator.handler.impl.AddedHandlerUtil - [bfd0af4a-a421-40e0-887c-2dc77083d5f6] METRIC: Running session for theia-cloud-demo
> 08:43:04.997 [pool-4-thread-1] DEBUG org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - Pinging sessions: [name=ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session version=22246 value=SessionSpec [name=ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session, appDefinition=theia-cloud-demo, user=foo@theia-cloud.io, url=35.242.207.84.sslip.io/instances/5f9d565d-d668-469e-ae2d-b05910aef0fb/, error=null, workspace=null, lastActivity=1681461745123]]
> 08:43:05.134 [pool-4-thread-1] INFO  org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - [ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session] Pinging session at 10.95.246.49
> 08:43:05.144 [pool-4-thread-1] INFO  org.eclipse.theia.cloud.operator.monitor.MonitorActivityTrackerImpl - [ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session] GET http://10.95.246.49:3000/monitor/activity/lastActivity
> 08:43:06.296 [pool-2-thread-1] TRACE org.eclipse.theia.cloud.operator.TheiaCloudImpl - [timeout-784153cb-8923-4764-b78a-ddfb23a70ed5] Session ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session will not be stopped automatically [NoTimout].
> 08:44:06.324 [pool-2-thread-1] TRACE org.eclipse.theia.cloud.operator.TheiaCloudImpl - [timeout-c9d26f8f-4c8e-43bf-ad0d-466914a04035] Session ws-1681461745070theia-cloud-demo-foo-theia-cloud-io-session will not be stopped automatically [NoTimout].
> ...
> no more pinging the session appears in the log
> ```
> 
> However both issues are more or less unrelated to the code in this PR and are more about hardening the monitor itself or improving the test setup with self signed certificates. Since this PR contains other valuable fixes, I think we should go ahead and merge this and create follow up tickets to harden the monitor.
> 
> So from my POV we can merge this after updating the versions in the package.json

Ah, I think I know whats wrong. Could it be that since we are using the same port for theia itself and the theia lastActivity extension, that the operator is redirected to the oauth2-proxy in case keycloak is enabled?
In the VSCode extension case we have a different port that is not guarded.

**Comment by sgraband** [comment]
> Ah, I think I know whats wrong. Could it be that since we are using the same port for theia itself and the theia lastActivity extension, that the operator is redirected to the oauth2-proxy in case keycloak is enabled?
In the VSCode extension case we have a different port that is not guarded.

Is all traffic on that port checked by the oauth2-proxy? Then this might be the problem you are experiencing. I am not to familiar with the proxy, but maybe we can ignore monitor paths on it? Or can the proxy be configured to correctly forward the monitor requests?

**Comment by jfaltermeier** [comment]
> > Ah, I think I know whats wrong. Could it be that since we are using the same port for theia itself and the theia lastActivity extension, that the operator is redirected to the oauth2-proxy in case keycloak is enabled?
> > In the VSCode extension case we have a different port that is not guarded.
> 
> Is all traffic on that port checked by the oauth2-proxy? Then this might be the problem you are experiencing. I am not to familiar with the proxy, but maybe we can ignore monitor paths on it? Or can the proxy be configured to correctly forward the monitor requests?
> 
> I will provide a commit with the version updates later today.

https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview
`--skip-auth-route` might be what we need to configure. I will create a dedicated follow up ticket for this.

**Review by jfaltermeier — APPROVED** [review]

**Comment by jfaltermeier** [comment]
Follow Up Tickets:
https://github.com/eclipsesource/theia-cloud/issues/164
https://github.com/eclipsesource/theia-cloud/issues/165
https://github.com/eclipsesource/theia-cloud/issues/166

**Pull request merged**